### PR TITLE
Support anonymized export on FHIR service

### PR DESCRIPTION
**Pull request by skywalker-wu**
## Description
In this PR, we add new feature of anonymized export in R4 and STU3. Customer can export anonymized data with customized anonymization configuration on destination storage.

## Related issues
Work item: [AB#74653](https://microsofthealth.visualstudio.com/f8da5110-49b1-4e9f-9022-2f58b6124ff9/_workitems/edit/74653)

## Testing
1. Tests for Anonymizer, ExportAnonymizerFactory, ExportDestinationArtifactProvider. 
2. Tests for ExportTask, for integration with different component and different FHIR service version.
2. E2E tests at API level for R4 and STU3. Also add E2E test cases for negative cases.

### CODE REVIEW

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+        private IExportClientInitializer<CloudBlobClient> _exportClientInitializer;
+        private ExportJobConfiguration _exportJobConfiguration;
+        private CloudBlobClient _blobClient;
+
+        public ExportDestinationArtifactProvider(
+            IExportClientInitializer<CloudBlobClient> exportClientInitializer,
+            IOptions<ExportJobConfiguration> exportJobConfiguration)
+        {
+            _exportClientInitializer = exportClientInitializer;
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:33]: Check for nulls `EnsureArg.IsNotNull(exportClientInitializer, nameof(exportClientInitializer))`

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            _exportJobConfiguration = exportJobConfiguration.Value;
+        }
+
+        public async Task FetchAsync(string blobNameWithETag, Stream targetStream, CancellationToken cancellationToken)
+        {
+            EnsureArg.IsNotNull(blobNameWithETag, nameof(blobNameWithETag));
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:39]: EnsureArg for `targetStream`

Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:39]: blobNameWithETag should probably use IsNotNullOrWhitespace

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+
+            string[] blobLocation = blobNameWithETag.Split(':', StringSplitOptions.RemoveEmptyEntries);
+            string blobName = blobLocation[0];
+            string eTag = blobLocation.Count() > 1 ? blobLocation[1] : null;
+
+            CloudBlobClient blobClient = await ConnectAsync(cancellationToken);
+            CloudBlobContainer container = blobClient.GetContainerReference(ANONYMIZATIONCONTAINER);
+            if (!await container.ExistsAsync(cancellationToken))
+            {
+                throw new FileNotFoundException(message: $"Container not found on the destination storage. {ANONYMIZATIONCONTAINER}");
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:49]: Move string to Resources

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+            }
+
+            CloudBlob blob = container.GetBlobReference(blobName);
+            if (await blob.ExistsAsync(cancellationToken))
+            {
+                if (CheckConfigurationIsTooLarge(blob))
+                {
+                    throw new AnonymizationConfigurationFetchException("Anonymization configuration is too large > 1MB.");
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:57]: Move string to Resources

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+            {
+                _blobClient = await _exportClientInitializer.GetAuthorizedClientAsync(_exportJobConfiguration, cancellationToken);
+            }
+
+            return _blobClient;
+        }
+
+        private bool CheckConfigurationIsTooLarge(CloudBlob blob)
+        {
+            return blob.Properties.Length > 8 * 1024 * 1024; // Max content length is 1 MB
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:97]: Isn't this 8mb?

File: src/Microsoft.Health.Fhir.Core/Features/KnownQueryParameterNames.cs [hunk 6]
```diff
@@ -55,5 +55,13 @@ public static class KnownQueryParameterNames
         public const string Id = "_id";
 
         public const string Type = "_type";
+
+        /// <summary>
+        /// The anonymization configuration
+        /// </summary>
+        /// <remarks>The anonymization configuration location and addition information. </remarks>
+        public const string AnonymizationConfigurationLocation = "_anonymization_config";
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/KnownQueryParameterNames.cs:63]: should these be camel case? e.g. `_lastUpdated`

File: src/Microsoft.Health.Fhir.Core/Features/Operations/IArtifactProvider.cs [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations
+{
+    public interface IArtifactProvider
+    {
+        Task FetchAsync(string location, Stream targetStream, CancellationToken cancellationToken);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/IArtifactProvider.cs:14]: Perhaps add documentation to this interface for future reference

File: src/Microsoft.Health.Fhir.R4.Core.UnitTests/Features/Export/ExportAnonymizerFactoryTests.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+	""fhirPathRules"": [
+		{
+			""path"": ""invalid()"",
+			""method"": ""redact""
+		}
+	]
+}";
+
+        [Fact]
+        public async Task GivenAValidAnonymizationConfiguration_WhenCreate_AnonymizerShouldBeCreated()
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.R4.Core.UnitTests/Features/Export/ExportAnonymizerFactoryTests.cs:43]: Perhaps `_WhenCreatingAnonymizer_`

File: src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizer.cs [hunk 9]
```diff
@@ -0,0 +1,28 @@
...
+        public ExportAnonymizer(AnonymizerEngine engine)
+        {
+            EnsureArg.IsNotNull(engine, nameof(engine));
+
+            _engine = engine;
+        }
+
+        public ResourceElement Anonymize(ResourceElement resourceElement)
+        {
+            return new ResourceElement(_engine.AnonymizeElement(resourceElement.Instance));
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizer.cs:25]: EnsureArg.IsNotNull resourceElement

File: src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+namespace Microsoft.Health.Fhir.Core.Features.Operations.Export
+{
+    public class ExportAnonymizerFactory : IAnonymizerFactory
+    {
+        private IArtifactProvider _artifactProvider;
+        private ILogger<ExportJobTask> _logger;
+
+        public ExportAnonymizerFactory(IArtifactProvider artifactProvider, ILogger<ExportJobTask> logger)
+        {
+            _artifactProvider = artifactProvider;
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs:22]: EnsureArgs

File: src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+            _logger = logger;
+        }
+
+        public async Task<IAnonymizer> CreateAnonymizerAsync(string configurationLocation, CancellationToken cancellationToken)
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs:26]: EnsureArgs

File: src/Microsoft.Health.Fhir.Shared.Api/Modules/OperationsModule.cs [hunk 12]
```diff
@@ -53,6 +53,8 @@ public void Load(IServiceCollection services)
             services.Add<ReindexJobWorker>()
                 .Singleton()
                 .AsSelf();
+
+            services.AddFactory<IScoped<IAnonymizer>>();
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Api/Modules/OperationsModule.cs:57]: Perhaps keep things together in the AnonymizerModule

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Operations/Export/ResourceToNdjsonBytesSerializer.cs [hunk 13]
```diff
@@ -31,13 +31,9 @@ public ResourceToNdjsonBytesSerializer(ResourceDeserializer resourceDeserializer
...
         /// <inheritdoc />
-        public byte[] Serialize(ResourceWrapper resourceWrapper)
+        public byte[] Serialize(ResourceElement resourceElement)
         {
-            EnsureArg.IsNotNull(resourceWrapper, nameof(resourceWrapper));
-
-            ResourceElement resource = _resourceDeserializer.DeserializeRaw(resourceWrapper.RawResource, resourceWrapper.Version, resourceWrapper.LastModified);
-
-            string resourceData = _fhirJsonSerializer.SerializeToString(resource.ToPoco<Resource>());
+            string resourceData = _fhirJsonSerializer.SerializeToString(resourceElement.ToPoco<Resource>());
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Operations/Export/ResourceToNdjsonBytesSerializer.cs:36]: EnsureArg

File: src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs [hunk 14]
```diff
@@ -0,0 +1,63 @@
...
+            {
+                try
+                {
+                    await _artifactProvider.FetchAsync(configurationLocation, stream, cancellationToken);
+                    stream.Position = 0;
+                }
+                catch (FileNotFoundException ex)
+                {
+                    _logger.LogError($"Anonymization configuration file not found: {configurationLocation}");
+                    throw new AnonymizationConfigurationNotFoundException(ex.Message, ex);
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs:38]: It seems that `$"Anonymization configuration file not found: {configurationLocation}"` should be the exception message. 
I think we've log the valuable message then throw an AnonymizationConfigurationNotFoundException with the original exception message which probably isn't as useful. Also given you're passing in the inner exception the original message will already be in there

File: src/Microsoft.Health.Fhir.Stu3.Core/Features/Operations/Export/ExportAnonymizerFactory.cs [hunk 15]
```diff
@@ -0,0 +1,63 @@
...
+using System;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Health.Fhir.Anonymizer.Core;
+
+namespace Microsoft.Health.Fhir.Core.Features.Operations.Export
+{
+    public class ExportAnonymizerFactory : IAnonymizerFactory
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Stu3.Core/Features/Operations/Export/ExportAnonymizerFactory.cs:15]: It doesn't seem ideal that we have all these files in two locations

File: src/Microsoft.Health.Fhir.Stu3.Core.UnitTests/Features/Export/ExportAnonymizerFactoryTests.cs [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+	""fhirPathRules"": [
+		{
+			""path"": ""invalid()"",
+			""method"": ""redact""
+		}
+	]
+}";
+
+        [Fact]
+        public async Task GivenAValidAnonymizationConfiguration_WhenCreate_AnonymizerShouldBeCreated()
```
Comment (brendankowitz) [inline src/Microsoft.Health.Fhir.Stu3.Core.UnitTests/Features/Export/ExportAnonymizerFactoryTests.cs:43]: As with the other file Perhaps _WhenCreatingAnonymizer_

File: test/Microsoft.Health.Fhir.R4.Tests.E2E/Rest/AnonymizedExportDataValidationTests.cs [hunk 17]
```diff
@@ -0,0 +1,256 @@
...
+}";
+
+        public AnonymizedExportDataValidationTests(HttpIntegrationTestFixture<StartupForAnonymizedExportTestProvider> fixture)
+        {
+            _testFhirClient = fixture.TestFhirClient;
+            _exportConfiguration = ((IOptions<ExportJobConfiguration>)(fixture.TestFhirServer as InProcTestFhirServer).Server.Services.GetService(typeof(IOptions<ExportJobConfiguration>))).Value;
+        }
+
+        [Fact]
+        public async Task GivenAValidConfigurationWithETag_WhenExportAnonymizedData_ResourceShouldBeAnonymized()
```
Comment (brendankowitz) [inline test/Microsoft.Health.Fhir.R4.Tests.E2E/Rest/AnonymizedExportDataValidationTests.cs:52]: _WhenExportingAnonymizedData_

File: test/Microsoft.Health.Fhir.Stu3.Tests.E2E/Rest/AnonymizedExportDataValidationTests.cs [hunk 18]
```diff
@@ -0,0 +1,256 @@
...
+}";
+
+        public AnonymizedExportDataValidationTests(HttpIntegrationTestFixture<StartupForAnonymizedExportTestProvider> fixture)
+        {
+            _testFhirClient = fixture.TestFhirClient;
+            _exportConfiguration = ((IOptions<ExportJobConfiguration>)(fixture.TestFhirServer as InProcTestFhirServer).Server.Services.GetService(typeof(IOptions<ExportJobConfiguration>))).Value;
+        }
+
+        [Fact]
+        public async Task GivenAValidConfigurationWithETag_WhenExportAnonymizedData_ResourceShouldBeAnonymized()
```
Comment (brendankowitz) [inline test/Microsoft.Health.Fhir.Stu3.Tests.E2E/Rest/AnonymizedExportDataValidationTests.cs:52]: _WhenExportingAnonymizedData_

File: src/Microsoft.Health.Fhir.Api/Configs/FeatureConfiguration.cs [hunk 19]
```diff
@@ -34,5 +34,10 @@ public class FeatureConfiguration
         /// Gets or sets a value indicating whether Validate is enabled or not.
         /// </summary>
         public bool SupportsValidate { get; set; }
+
+        /// <summary>
+        /// Gets or sets a value indicating whether anonymized export is enabled or not.
+        /// </summary>
+        public bool SupportAnonymizedExport { get; set; }
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Api/Configs/FeatureConfiguration.cs:41]: Should this be an option of the ExportJobConfiguration?

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/AnonymizationConfigurationFetchException.cs [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+                OperationOutcomeConstants.IssueType.Exception,
+                message));
+        }
+
+        public AnonymizationConfigurationFetchException(string message, Exception exception)
+            : base(message, exception)
+        {
+            Issues.Add(new Models.OperationOutcomeIssue(
+                OperationOutcomeConstants.IssueSeverity.Error,
+                OperationOutcomeConstants.IssueType.NotFound,
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/AnonymizationConfigurationFetchException.cs:33]: Why is this one a NotFound issue type and the one above is an Exception issue type?

File: src/Microsoft.Health.Fhir.Shared.Api/Controllers/ExportController.cs [hunk 21]
```diff
@@ -50,36 +51,46 @@ public class ExportController : Controller
...
         [Route(KnownRoutes.Export)]
         [ServiceFilter(typeof(ValidateExportRequestFilterAttribute))]
         [AuditEventType(AuditEventSubType.Export)]
-        public async Task<IActionResult> Export([FromQuery(Name = KnownQueryParameterNames.Since)] PartialDateTime since, [FromQuery(Name = KnownQueryParameterNames.Type)] string resourceType)
+        public async Task<IActionResult> Export([FromQuery(Name = KnownQueryParameterNames.Since)] PartialDateTime since, [FromQuery(Name = KnownQueryParameterNames.Type)] string resourceType, [FromQuery(Name = KnownQueryParameterNames.AnonymizationConfigurationLocation)] string anonymizationConfigLocation, [FromQuery(Name = KnownQueryParameterNames.AnonymizationConfigurationFileEtag)] string anonymizationConfigFileETag)
         {
             CheckIfExportIsEnabled();
-            return await SendExportRequest(ExportJobType.All, since, resourceType);
+
+            if (!string.IsNullOrWhiteSpace(anonymizationConfigLocation))
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Shared.Api/Controllers/ExportController.cs:88]: Should this also check if the anonymizationConfigFileETag is provided?

File: src/Microsoft.Health.Fhir.Shared.Core/Features/Operations/Export/ResourceToNdjsonBytesSerializer.cs [hunk 22]
```diff
@@ -31,13 +31,9 @@ public ResourceToNdjsonBytesSerializer(ResourceDeserializer resourceDeserializer
         }
 
         /// <inheritdoc />
-        public byte[] Serialize(ResourceWrapper resourceWrapper)
+        public byte[] Serialize(ResourceElement resourceElement)
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Shared.Core/Features/Operations/Export/ResourceToNdjsonBytesSerializer.cs:34]: The ResourceDeserializer is no longer needed in the constructor of this class.

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs [hunk 23]
```diff
@@ -263,7 +292,18 @@ private async Task UpdateJobRecordAsync(CancellationToken cancellationToken)
                     }
                 }
 
-                await ProcessSearchResultsAsync(searchResult.Results, currentBatchId, cancellationToken);
+                if (IsAnonymizedExportJob())
+                {
+                    string configurationWithEtag =
+                        $"{_exportJobRecord.AnonymizationConfigurationLocation}:{_exportJobRecord.AnonymizationConfigurationFileETag}";
+
+                    IAnonymizer anonymizer = await _anonymizerFactory.Value.CreateAnonymizerAsync(configurationWithEtag, cancellationToken);
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/ExportJobTask.cs:300]: Could the anonymizer be created outside the while loop? That would avoid recreating it for every page of results.

File: src/Microsoft.Health.Fhir.Api/Features/Filters/ValidateExportRequestFilterAttribute.cs [hunk 24]
```diff
@@ -32,6 +32,8 @@ public ValidateExportRequestFilterAttribute()
             _supportedQueryParams = new HashSet<string>(StringComparer.Ordinal)
             {
                 KnownQueryParameterNames.Since,
+                KnownQueryParameterNames.AnonymizationConfigurationLocation,
+                KnownQueryParameterNames.AnonymizationConfigurationFileEtag,
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Api/Features/Filters/ValidateExportRequestFilterAttribute.cs:36]: These query params are only supported for the base $export endpoint, not the patient or group endpoints.

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/IResourceToByteArraySerializer.cs [hunk 25]
```diff
@@ -15,8 +16,8 @@ public interface IResourceToByteArraySerializer
         /// <summary>
         /// Serializes the resource represented by <see cref="ResourceWrapper"/> to byte array.
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/IResourceToByteArraySerializer.cs:17]: Update documentation

File: src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs [hunk 26]
```diff
@@ -0,0 +1,63 @@
...
+                {
+                    string configurationContent = await reader.ReadToEndAsync();
+                    try
+                    {
+                        var engine = new AnonymizerEngine(AnonymizerConfigurationManager.CreateFromSettingsInJson(configurationContent));
+                        return new ExportAnonymizer(engine);
+                    }
+                    catch (Exception ex)
+                    {
+                        _logger.LogError($"Loca anonymization configuration file failed: {ex.Message}");
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs:56]: Typo

File: src/Microsoft.Health.Fhir.Stu3.Core/Features/Operations/Export/ExportAnonymizerFactory.cs (hunk 15 above)
Comment (LTA-Thinking) [reply src/Microsoft.Health.Fhir.Stu3.Core/Features/Operations/Export/ExportAnonymizerFactory.cs:15]: If the goal is to not support R5, would it be simpler to detect the FHIR version in the AnonymizationModule and not add the services if it is R5?

File: src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs [hunk 27]
```diff
@@ -283,6 +283,19 @@ public async Task<Uri> ExportAsync(string path = "", string parameters = "", Can
             return response.Content.Headers.ContentLocation;
         }
 
+        public async Task<Uri> AnonymizedExportAsync(string anonymizationConfig, string etag = null, CancellationToken cancellationToken = default)
+        {
+            using var message = new HttpRequestMessage(HttpMethod.Get, $"$export?_anonymization_config={anonymizationConfig}&_anonymization_config_etag={etag}");
```
Comment (cunninghamjc) [inline src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs:288]: Do the query string parameter arguments need to be URL encoded first?

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs (hunk 2 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:39]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs (hunk 1 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:33]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs (hunk 3 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:53]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs (hunk 4 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:61]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Core/Features/KnownQueryParameterNames.cs (hunk 6 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Core/Features/KnownQueryParameterNames.cs:63]: Yes, updated to follow the same convention. Thanks

File: src/Microsoft.Health.Fhir.Shared.Api/Modules/OperationsModule.cs (hunk 12 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Shared.Api/Modules/OperationsModule.cs:57]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Stu3.Core/Features/Operations/Export/ExportAnonymizerFactory.cs (hunk 15 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Stu3.Core/Features/Operations/Export/ExportAnonymizerFactory.cs:16]: Check the version in code not work for this scénario, it would failed at compile code. We do not have R5 Nuget now. 

I updated the PR with below change to work around this issue before R5 anonymizer release:

1. Remove code file from shared.*..projitems
2. Add None reference in shared.*.shproj
3. Add file link reference in R4 and STU3 project.

Not sure any concern for this solution? Thanks!

File: src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs (hunk 26 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Api/Configs/FeatureConfiguration.cs (hunk 19 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Api/Configs/FeatureConfiguration.cs]: We would prefer default support anonymized export in FHIR server, so we do not want to make it as a settings for export. We put the flag here is to make sure it would not enable by default during preview stage and not want to enable it on PaaS automatically after OSS nuegt updated.

File: src/Microsoft.Health.Fhir.Stu3.Core.UnitTests/Features/Export/ExportAnonymizerFactoryTests.cs (hunk 16 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Stu3.Core.UnitTests/Features/Export/ExportAnonymizerFactoryTests.cs]: Thanks, updated.

File: test/Microsoft.Health.Fhir.R4.Tests.E2E/Rest/AnonymizedExportDataValidationTests.cs (hunk 17 above)
Comment (skywalker-wu) [reply test/Microsoft.Health.Fhir.R4.Tests.E2E/Rest/AnonymizedExportDataValidationTests.cs]: Thanks, updated.

File: test/Microsoft.Health.Fhir.Stu3.Tests.E2E/Rest/AnonymizedExportDataValidationTests.cs (hunk 18 above)
Comment (skywalker-wu) [reply test/Microsoft.Health.Fhir.Stu3.Tests.E2E/Rest/AnonymizedExportDataValidationTests.cs]: Thanks, updated.

File: src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs (hunk 14 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.R4.Core/Features/Operations/Export/ExportAnonymizerFactory.cs]: Yes, I already put these information in inner exception. Updated.

File: src/Microsoft.Health.Fhir.R4.Core.UnitTests/Features/Export/ExportAnonymizerFactoryTests.cs (hunk 8 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.R4.Core.UnitTests/Features/Export/ExportAnonymizerFactoryTests.cs]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Core/Features/Operations/IArtifactProvider.cs (hunk 7 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/IArtifactProvider.cs:14]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs (hunk 27 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs]: it support both original string and encoded url.

File: src/Microsoft.Health.Fhir.Api/Features/Filters/ValidateExportRequestFilterAttribute.cs (hunk 24 above)
Comment (LTA-Thinking) [reply src/Microsoft.Health.Fhir.Api/Features/Filters/ValidateExportRequestFilterAttribute.cs:36]: I think having a special case for the default export would be good as a user could enter these parameters for the other export endpoints and get no indication that they were not honored. You should be able to use context.HttpContext.Request.Path to figure out which endpoint the request is being sent to.

File: src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs [hunk 28]
```diff
@@ -1454,6 +1476,243 @@ public async Task GivenAGroupExportJobWithANonExistantGroup_WhenExecuted_ThenThe
             Assert.Equal("test", _exportJobRecord.FailureDetails.FailureReason);
         }
 
+        [Fact]
+        public async Task GivenAnonymizedExportJob_WhenExecuted_ThenItShouldExportAllAnonymizedResources()
+        {
+            bool capturedSearch = false;
+
+            var exportJobRecordWithOneResource = new ExportJobRecord(
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs:1484]: There is a helper method (CreateExportJobRecord) that can save you some lines of code here and in the other new tests.

File: src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs [hunk 29]
```diff
@@ -1454,6 +1476,243 @@ public async Task GivenAGroupExportJobWithANonExistantGroup_WhenExecuted_ThenThe
...
+            // First search should not have continuation token in the list of query parameters.
+            _searchService.SearchAsync(
+                Arg.Any<string>(),
+                Arg.Any<IReadOnlyList<Tuple<string, string>>>(),
+                _cancellationToken)
+                .Returns(x =>
+                {
+                    return CreateSearchResult(new[]
+                        {
+                            new SearchResultEntry(
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs:1506]: There is a helper method (CreateSearchResultEntry) that could be used here and in other tests.

File: src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs [hunk 30]
```diff
@@ -1454,6 +1476,243 @@ public async Task GivenAGroupExportJobWithANonExistantGroup_WhenExecuted_ThenThe
...
+            await anonymizedExportJobTask.ExecuteAsync(_exportJobRecord, _weakETag, _cancellationToken);
+
+            string exportedValue = inMemoryDestinationClient.GetExportedData(new Uri(PatientFileName, UriKind.Relative));
+
+            Assert.Equal("anonymized-resource", exportedValue);
+            Assert.True(capturedSearch);
+        }
+
+        [Fact]
+        public async Task GivenAnonymizationConfigurationNotExist_WhenExecuted_ThenJobStatusShouldBeUpdatedToFailed()
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs:1553]: Could some of these anonymizer exception tests be combined into a Theory with InlineData to give the exception type?

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs [hunk 31]
```diff
@@ -0,0 +1,100 @@
...
+using Microsoft.Extensions.Options;
+using Microsoft.Health.Fhir.Core.Configs;
+using Microsoft.Health.Fhir.Core.Features.Operations;
+using Microsoft.Health.Fhir.Core.Features.Operations.Export.ExportDestinationClient;
+
+namespace Microsoft.Health.Fhir.Azure
+{
+    public class ExportDestinationArtifactProvider : IArtifactProvider
+    {
+        private const string ANONYMIZATIONCONTAINER = "anonymization";
```
Comment (cunninghamjc) [inline src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:24]: Is it our convention to use all uppercase for const variable names?

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs (hunk 5 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs]: yes, updated to 1 MB. Thanks

File: src/Microsoft.Health.Fhir.Api/Features/Filters/ValidateExportRequestFilterAttribute.cs (hunk 24 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Api/Features/Filters/ValidateExportRequestFilterAttribute.cs:36]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs (hunk 29 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs:1506]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs (hunk 28 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs:1484]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs (hunk 30 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Core.UnitTests/Features/Operations/Export/ExportJobTaskTests.cs:1553]: Thanks, updated.

File: src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs (hunk 31 above)
Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Azure/ExportDestinationArtifactProvider.cs:24]: Thanks, I should keep same convention as others. Updated.

File: src/Microsoft.Health.Fhir.Core/Features/Operations/Export/AnonymizationConfigurationNotFoundException.cs [hunk 32]
```diff
@@ -0,0 +1,37 @@
...
+                OperationOutcomeConstants.IssueType.Exception,
+                message));
+        }
+
+        public AnonymizationConfigurationNotFoundException(string message, Exception exception)
+            : base(message, exception)
+        {
+            Issues.Add(new Models.OperationOutcomeIssue(
+                OperationOutcomeConstants.IssueSeverity.Error,
+                OperationOutcomeConstants.IssueType.NotFound,
```
Comment (LTA-Thinking) [inline src/Microsoft.Health.Fhir.Core/Features/Operations/Export/AnonymizationConfigurationNotFoundException.cs:33]: Mismatched issue type here as well.

Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Core/Features/Operations/Export/AnonymizationConfigurationNotFoundException.cs:33]: Sure, updated.

File: src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs [hunk 33]
```diff
@@ -283,6 +283,19 @@ public async Task<Uri> ExportAsync(string path = "", string parameters = "", Can
             return response.Content.Headers.ContentLocation;
         }
 
+        public async Task<Uri> AnonymizedExportAsync(string anonymizationConfig, string etag = null, CancellationToken cancellationToken = default)
+        {
+            using var message = new HttpRequestMessage(HttpMethod.Get, $"$export?_anonymizationConfig={anonymizationConfig}&_anonymizationConfigEtag={etag}");
```
Comment (cunninghamjc) [inline src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs:288]: Do we need to URL encode the query parameter values?

Comment (skywalker-wu) [reply src/Microsoft.Health.Fhir.Shared.Client/FhirClient.cs]: Thanks, add encode to be safe here.
